pkg/backends/calico/routes.go

```diff
@@ -14,31 +14,299 @@
-	routeString := os.Getenv("CALICO_STATIC_ROUTES")
-	cidrs := []string{}
+	var (
+		isDynamic   bool
+		ch          = make(chan struct{})
+		cidrs       = []string{}
+		routeString = os.Getenv(envStaticRoutes)
+	)
+
+	// parse routeString
 	for _, route := range strings.Split(routeString, ",") {
 		cidr := strings.TrimSpace(route)
-		if cidr != "" {
+		if len(cidr) == 0 {
+			// skip empty string
+			continue
+		} else if cidr == clusterIPStr {
+			// handle k8s-cluster-ips
+			isDynamic = true
+		} else {
+			// consider anything else as cidr
+			if _, _, err := net.ParseCIDR(cidr); err != nil {
+				log.WithField("cidr", cidr).WithError(err).Warn("Start: failed to parse cidr, passing")
+				continue
+			}
 			cidrs = append(cidrs, cidr)
 		}
 	}
-	rg.client.updateRoutes(cidrs)
+
+	// affix cidrs to the route map
+	rg.svcRouteMap[staticRoutesKey] = cidrs
+	rg.updateRoutes()
+
+	// start the listeners
+	if isDynamic {
+		go rg.svcInformer.Run(ch)
+		go rg.epInformer.Run(ch)
+	}
+
 	return
 }
+
+// getServiceForEndpoints retrieves the corresponding svc for the given ep
+func (rg *routeGenerator) getServiceForEndpoints(ep *v1.Endpoints) (*v1.Service, string) {
+	// get key
+	key, err := cache.MetaNamespaceKeyFunc(ep)
+	if err != nil {
+		log.WithField("ep", ep.Name).WithError(err).Warn("getServiceForEndpoints: error on retrieving key for endpoint, passing")
+		return nil, ""
+	}
+	// get svc
+	svcIface, exists, err := rg.svcIndexer.GetByKey(key)
+	if err != nil {
+		log.WithField("key", key).WithError(err).Warn("getServiceForEndpoints: error on retrieving service for key, passing")
+		return nil, key
+	} else if !exists {
+		log.WithField("key", key).Warn("getServiceForEndpoints: service for key not found, passing")
+		return nil, key
+	}
+	return svcIface.(*v1.Service), key
+}
+
+// getEndpointsForService retrieves the corresponding ep for the given svc
+func (rg *routeGenerator) getEndpointsForService(svc *v1.Service) (*v1.Endpoints, string) {
+	// get key
+	key, err := cache.MetaNamespaceKeyFunc(svc)
+	if err != nil {
+		log.WithField("svc", svc.Name).WithError(err).Warn("getEndpointsForService: error on retrieving key for service, passing")
+		return nil, ""
+	}
+	// get ep
+	epIface, exists, err := rg.epIndexer.GetByKey(key)
+	if err != nil {
+		log.WithField("key", key).WithError(err).Warn("getEndpointsForService: error on retrieving endpoint for key, passing")
+		return nil, key
+	} else if !exists {
+		log.WithField("key", key).Warn("getEndpointsForService: service for endpoint not found, passing")
+		return nil, key
+	}
+	return epIface.(*v1.Endpoints), key
+}
+
+// setRouteForSvc handles the main logic to check if a specified service or endpoint
+// should have its route advertised by the node running this code
+func (rg *routeGenerator) setRouteForSvc(svc *v1.Service, ep *v1.Endpoints) {
+
+	// ensure both are not nil
+	if svc == nil && ep == nil {
+		log.Warn("setRouteForSvc: both service and endpoint cannot be nil, passing...")
```

Comment:
nit: this should be an Error log - we should never hit this branch unless our code is bugged.